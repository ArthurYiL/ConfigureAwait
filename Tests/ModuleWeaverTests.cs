using System;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Runtime.Versioning;
using ApprovalTests.Namers;
using Fody;
using Xunit.Abstractions;

public partial class ModuleWeaverTests
{
    static TestResult testResult;
    IDisposable disposable;

    static ModuleWeaverTests()
    {
        Assembly.Load("xunit.assert");
        var weavingTask = new ModuleWeaver();
        testResult = weavingTask.ExecuteTestRun("AssemblyToProcess.dll");
    }

    public ModuleWeaverTests(ITestOutputHelper output) :
        base(output)
    {
        var targetFrameworkAttribute = (TargetFrameworkAttribute)Assembly.GetExecutingAssembly()
            .GetCustomAttributes(typeof(TargetFrameworkAttribute), false)
            .SingleOrDefault();
        var framework = targetFrameworkAttribute.FrameworkDisplayName.Split(' ').Last();
#if DEBUG
        var suffix = "Debug" + framework;
#else
        var suffix = "Release" + framework;
#endif 
        disposable = NamerFactory.AsEnvironmentSpecificTest(() => suffix);
    }

    public override void Dispose()
    {
        base.Dispose();
        disposable.Dispose();
    }
}