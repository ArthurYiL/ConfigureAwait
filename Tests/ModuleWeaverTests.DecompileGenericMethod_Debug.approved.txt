.class private auto ansi sealed beforefieldinit AssemblyToProcess.GenericMethod
extends [System.Runtime]System.Object
{
.class auto ansi sealed nested private beforefieldinit '<Method>d__0`1'<TItem>
extends [System.Runtime]System.Object
implements [System.Threading.Tasks]System.Runtime.CompilerServices.IAsyncStateMachine
{
.field public int32 '<>1__state'
.field public valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder '<>t__builder'
.field public class [System.Runtime]System.Threading.Tasks.Task`1<!TItem> itemTask
.field public class AssemblyToProcess.GenericMethod '<>4__this'
.field private !TItem '<item>5__1'
.field private !TItem '<>s__2'
.field private valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem> '<>u__1'
.method public hidebysig specialname rtspecialname
instance void  .ctor() cil managed
{
.maxstack  8
IL_0000:  ldarg.0
IL_0001:  call       instance void [System.Runtime]System.Object::.ctor()
IL_0006:  nop
IL_0007:  ret
}
.method private hidebysig newslot virtual final
instance void  MoveNext() cil managed
{
.override [System.Threading.Tasks]System.Runtime.CompilerServices.IAsyncStateMachine::MoveNext
.maxstack  3
.locals init (int32 V_0,
valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem> V_1,
valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1<!TItem> V_2,
class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem> V_3,
class [System.Runtime]System.Exception V_4)
IL_0000:  ldarg.0
IL_0001:  ldfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>1__state'
IL_0006:  stloc.0
.try
{
IL_0007:  ldloc.0
IL_0008:  brfalse.s  IL_000c
IL_000a:  br.s       IL_000e
IL_000c:  br.s       IL_0051
IL_000e:  nop
IL_000f:  ldarg.0
IL_0010:  ldfld      class [System.Runtime]System.Threading.Tasks.Task`1<!0> class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::itemTask
IL_0015:  ldc.i4.0
IL_0016:  callvirt   instance valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1<!0> class [netstandard]System.Threading.Tasks.Task`1<!TItem>::ConfigureAwait(bool)
IL_001b:  stloc.2
IL_001c:  ldloca.s   V_2
IL_001e:  call       instance valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!0> valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1<!TItem>::GetAwaiter()
IL_0023:  stloc.1
IL_0024:  ldloca.s   V_1
IL_0026:  call       instance bool valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem>::get_IsCompleted()
IL_002b:  brtrue.s   IL_006d
IL_002d:  ldarg.0
IL_002e:  ldc.i4.0
IL_002f:  dup
IL_0030:  stloc.0
IL_0031:  stfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>1__state'
IL_0036:  ldarg.0
IL_0037:  ldloc.1
IL_0038:  stfld      valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!0> class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>u__1'
IL_003d:  ldarg.0
IL_003e:  stloc.3
IL_003f:  ldarg.0
IL_0040:  ldflda     valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>t__builder'
IL_0045:  ldloca.s   V_1
IL_0047:  ldloca.s   V_3
IL_0049:  call       instance void [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::AwaitUnsafeOnCompleted<valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem>,class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>>(!!0&,
!!1&)
IL_004e:  nop
IL_004f:  leave.s    IL_00c2
IL_0051:  ldarg.0
IL_0052:  ldfld      valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!0> class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>u__1'
IL_0057:  stloc.1
IL_0058:  ldarg.0
IL_0059:  ldflda     valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!0> class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>u__1'
IL_005e:  initobj    valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem>
IL_0064:  ldarg.0
IL_0065:  ldc.i4.m1
IL_0066:  dup
IL_0067:  stloc.0
IL_0068:  stfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>1__state'
IL_006d:  ldarg.0
IL_006e:  ldloca.s   V_1
IL_0070:  call       instance !0 valuetype [netstandard]System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1/ConfiguredTaskAwaiter<!TItem>::GetResult()
IL_0075:  stfld      !0 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>s__2'
IL_007a:  ldarg.0
IL_007b:  ldarg.0
IL_007c:  ldfld      !0 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>s__2'
IL_0081:  stfld      !0 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<item>5__1'
IL_0086:  ldarg.0
IL_0087:  ldflda     !0 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>s__2'
IL_008c:  initobj    !TItem
IL_0092:  leave.s    IL_00ae
}  // end .try
catch [System.Runtime]System.Exception
{
IL_0094:  stloc.s    V_4
IL_0096:  ldarg.0
IL_0097:  ldc.i4.s   -2
IL_0099:  stfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>1__state'
IL_009e:  ldarg.0
IL_009f:  ldflda     valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>t__builder'
IL_00a4:  ldloc.s    V_4
IL_00a6:  call       instance void [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetException(class [System.Runtime]System.Exception)
IL_00ab:  nop
IL_00ac:  leave.s    IL_00c2
}  // end handler
IL_00ae:  ldarg.0
IL_00af:  ldc.i4.s   -2
IL_00b1:  stfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>1__state'
IL_00b6:  ldarg.0
IL_00b7:  ldflda     valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!TItem>::'<>t__builder'
IL_00bc:  call       instance void [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetResult()
IL_00c1:  nop
IL_00c2:  ret
}
.method private hidebysig newslot virtual final
instance void  SetStateMachine(class [System.Threading.Tasks]System.Runtime.CompilerServices.IAsyncStateMachine stateMachine) cil managed
{
.override [System.Threading.Tasks]System.Runtime.CompilerServices.IAsyncStateMachine::SetStateMachine
.maxstack  8
IL_0000:  ret
}
}
.method public hidebysig instance class [System.Runtime]System.Threading.Tasks.Task
Method<TItem>(class [System.Runtime]System.Threading.Tasks.Task`1<!!TItem> itemTask) cil managed
{
63 65 73 73 2E 47 65 6E 65 72 69 63 4D 65 74 68   // cess.GenericMeth
6F 64 2B 3C 4D 65 74 68 6F 64 3E 64 5F 5F 30 60   // od+<Method>d__0`
31 00 00 )                                        // 1..
.maxstack  2
.locals init (class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem> V_0,
valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder V_1)
IL_0000:  newobj     instance void class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::.ctor()
IL_0005:  stloc.0
IL_0006:  ldloc.0
IL_0007:  ldarg.0
IL_0008:  stfld      class AssemblyToProcess.GenericMethod class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::'<>4__this'
IL_000d:  ldloc.0
IL_000e:  ldarg.1
IL_000f:  stfld      class [System.Runtime]System.Threading.Tasks.Task`1<!0> class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::itemTask
IL_0014:  ldloc.0
IL_0015:  call       valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::Create()
IL_001a:  stfld      valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::'<>t__builder'
IL_001f:  ldloc.0
IL_0020:  ldc.i4.m1
IL_0021:  stfld      int32 class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::'<>1__state'
IL_0026:  ldloc.0
IL_0027:  ldfld      valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::'<>t__builder'
IL_002c:  stloc.1
IL_002d:  ldloca.s   V_1
IL_002f:  ldloca.s   V_0
IL_0031:  call       instance void [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::Start<class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!0>>(!!0&)
IL_0036:  ldloc.0
IL_0037:  ldflda     valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder class AssemblyToProcess.GenericMethod/'<Method>d__0`1'<!!TItem>::'<>t__builder'
IL_003c:  call       instance class [System.Runtime]System.Threading.Tasks.Task [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::get_Task()
IL_0041:  ret
}
.method public hidebysig specialname rtspecialname
instance void  .ctor() cil managed
{
.maxstack  8
IL_0000:  ldarg.0
IL_0001:  call       instance void [System.Runtime]System.Object::.ctor()
IL_0006:  nop
IL_0007:  ret
}
}
